File: pipeline/test/peer-sync.test.mjs
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { test } from 'node:test';
import { fileURLToPath } from 'node:url';

import { buildRequest, computeDiff } from '../lib/peer-sync.mjs';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, '../..');

function loadFixture(name) {
  return JSON.parse(readFileSync(path.join(ROOT, 'fixtures', name), 'utf8'));
}

const exchange = loadFixture('protocol-exchange-v0.json');
const peerA = loadFixture(exchange.peer_summaries['node-a']);
const peerB = loadFixture(exchange.peer_summaries['node-b']);

const diffMessage = exchange.messages.find((message) => message.type === 'DIFF');
const requestMessage = exchange.messages.find((message) => message.type === 'REQUEST');

test('computeDiff finds the chunk node-a is missing from node-b', () => {
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });

  assert.equal(diff.manifest_id, diffMessage.manifest_id);
  assert.deepEqual(diff.missing_chunks, diffMessage.missing_chunks);
  assert.deepEqual(diff.stale_chunks, diffMessage.stale_chunks);

  assert.deepEqual(
    diff.missing_chunks.map((chunk) => chunk.chunk_id),
    exchange.expected.a_missing_chunks_before_sync,
  );
});

test('buildRequest matches the fixture REQUEST message', () => {
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });
  const request = buildRequest(diff, { resume: requestMessage.resume });

  assert.deepEqual(request.chunks, requestMessage.chunks);
  assert.equal(request.max_total_bytes, requestMessage.max_total_bytes);
  assert.equal(request.resume, requestMessage.resume);
});

test('after applying the request, node-a is missing nothing', () => {
  // Simulate APPLY: node-a merges every requested chunk into its own summary.
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });
  const merged = new Map(
    peerA.datasets[0].chunks.map((chunk) => [chunk.chunk_id, chunk]),
  );
  for (const chunk of diff.missing_chunks) {
    merged.set(chunk.chunk_id, chunk);
  }

  const peerAAfterSync = {
    ...peerA,
    datasets: [{ ...peerA.datasets[0], chunks: [...merged.values()] }],
  };
  const diffAfterSync = computeDiff(peerAAfterSync, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });

  assert.deepEqual(
    diffAfterSync.missing_chunks.map((chunk) => chunk.chunk_id),
    exchange.expected.a_missing_chunks_after_sync,
  );
});

test('computeDiff rejects two manifests that both claim the same dataset_version', () => {
  const peerBWrongManifest = {
    ...peerB,
    datasets: [{ ...peerB.datasets[0], manifest_id: 'demo:official:999' }],
  };

  assert.throws(
    () => computeDiff(peerA, peerBWrongManifest, {
      datasetId: diffMessage.dataset_id,
      namespace: diffMessage.namespace,
    }),
    /manifest mismatch/,
  );
});

test('computeDiff treats a peer carrying a newer manifest_id as a DTN supersession, not an error', () => {
  // A walks into a shelter carrying v137; everyone there (peerB) is still on
  // v136. This is the exact scenario opportunistic contact exists for, so it
  // must produce a REQUEST instead of throwing.
  const peerNewerManifest = {
    ...peerA,
    datasets: [{
      ...peerA.datasets[0],
      manifest_id: 'resilientgeo-demo:manifest:137',
      manifest_hash: 'sha256:eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
      dataset_version: 137,
      chunks: [
        {
          chunk_id: 'resilientgeo-demo:chunk:137:dahu:road:000',
          chunk_hash: 'sha256:1111111111111111111111111111111111111111111111111111111111111111',
          size_bytes: 512,
          priority: 'CRITICAL',
          state: 'available',
        },
      ],
    }],
  };

  const diff = computeDiff(peerNewerManifest, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });

  // Remote (peerB, v136) has nothing local doesn't already have for this
  // dataset -- local's own newer manifest_id (v137) stays current.
  assert.equal(diff.manifest_id, 'resilientgeo-demo:manifest:137');
  assert.equal(diff.superseded_manifest_id, null);
  assert.deepEqual(diff.missing_chunks, []);
  assert.deepEqual(diff.stale_chunks, []);
});

test('computeDiff requests every chunk from a peer carrying a newer manifest_id', () => {
  const peerBNewerManifest = {
    ...peerB,
    datasets: [{
      ...peerB.datasets[0],
      manifest_id: 'resilientgeo-demo:manifest:137',
      manifest_hash: 'sha256:eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
      dataset_version: 137,
      chunks: [
        {
          chunk_id: 'resilientgeo-demo:chunk:137:dahu:road:000',
          chunk_hash: 'sha256:1111111111111111111111111111111111111111111111111111111111111111',
          size_bytes: 512,
          priority: 'CRITICAL',
          state: 'available',
        },
      ],
    }],
  };

  const diff = computeDiff(peerA, peerBNewerManifest, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });

  assert.equal(diff.manifest_id, 'resilientgeo-demo:manifest:137');
  assert.equal(diff.superseded_manifest_id, 'resilientgeo-demo:manifest:136');
  assert.deepEqual(diff.stale_chunks, []);
  assert.deepEqual(
    diff.missing_chunks.map((chunk) => chunk.chunk_id),
    ['resilientgeo-demo:chunk:137:dahu:road:000'],
  );

  const request = buildRequest(diff);
  assert.equal(request.superseded_manifest_id, 'resilientgeo-demo:manifest:136');
});

test('buildRequest resumes from a per-chunk byte offset instead of restarting at 0', () => {
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });
  const target = diff.missing_chunks[0];

  const fresh = buildRequest(diff);
  assert.equal(fresh.chunks[0].offset_bytes, 0);
  assert.equal(fresh.chunks[0].max_bytes, target.size_bytes);

  // Same DIFF, but this node already holds the first 400 bytes from an
  // earlier contact that was cut short. Without this, every contact would
  // re-request the whole chunk and never finish it at BLE throughput.
  const resumed = buildRequest(diff, { offsets: { [target.chunk_id]: 400 } });
  assert.equal(resumed.chunks[0].offset_bytes, 400);
  assert.equal(resumed.chunks[0].max_bytes, target.size_bytes - 400);
  assert.equal(resumed.max_total_bytes, target.size_bytes - 400);
});

test('buildRequest drops a chunk that is already fully held', () => {
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });
  const target = diff.missing_chunks[0];

  const request = buildRequest(diff, { offsets: { [target.chunk_id]: target.size_bytes } });
  assert.deepEqual(request.chunks.map((chunk) => chunk.chunk_id), []);
  assert.equal(request.max_total_bytes, 0);
});

test('buildRequest rejects an offset past the end of the chunk', () => {
  const diff = computeDiff(peerA, peerB, {
    datasetId: diffMessage.dataset_id,
    namespace: diffMessage.namespace,
  });
  const target = diff.missing_chunks[0];

  assert.throws(
    () => buildRequest(diff, { offsets: { [target.chunk_id]: target.size_bytes + 1 } }),
    RangeError,
  );
  assert.throws(() => buildRequest(diff, { offsets: { [target.chunk_id]: -1 } }), RangeError);
});

File: android/app/src/test/java/com/resilientgeo/mesh/protocol/PeerSyncTest.kt
package com.resilientgeo.mesh.protocol

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Test

/**
 * Ports every case in `pipeline/test/peer-sync.test.mjs` to prove
 * `PeerSync` behaves identically to the Node module it was ported from,
 * against the exact same fixtures (`fixtures/protocol-exchange-v0.json`,
 * `peer-a/b-summary-v0.json`). No device or emulator needed — this is the
 * one deliverable 甲 is waiting on before wiring real transfers into
 * `BleGattTransport`.
 */
class PeerSyncTest {

    private val exchange = PeerSyncTestFixtures.exchange()
    private val peerA = PeerSyncTestFixtures.peerSummary("peer-a-summary-v0.json")
    private val peerB = PeerSyncTestFixtures.peerSummary("peer-b-summary-v0.json")

    private val diffMessage = findMessage("DIFF")
    private val requestMessage = findMessage("REQUEST")
    private val datasetId = diffMessage.getString("dataset_id")
    private val namespace = diffMessage.getString("namespace")

    private fun findMessage(type: String): JSONObject {
        val messages = exchange.getJSONArray("messages")
        for (i in 0 until messages.length()) {
            val message = messages.getJSONObject(i)
            if (message.getString("type") == type) return message
        }
        error("fixture has no $type message")
    }

    private fun chunkIdsOf(array: org.json.JSONArray): List<String> =
        (0 until array.length()).map { array.getJSONObject(it).getString("chunk_id") }

    @Test
    fun `computeDiff finds the chunk node-a is missing from node-b`() {
        val diff = PeerSync.computeDiff(peerA, peerB, datasetId, namespace)

        assertEquals(diffMessage.getString("manifest_id"), diff.manifestId)
        assertEquals(chunkIdsOf(diffMessage.getJSONArray("missing_chunks")), diff.missingChunks.map { it.chunkId })
        assertEquals(0, diffMessage.getJSONArray("stale_chunks").length())
        assertEquals(emptyList<String>(), diff.staleChunks.map { it.chunkId })

        val expectedBeforeSync = exchange.getJSONObject("expected").getJSONArray("a_missing_chunks_before_sync")
        assertEquals(
            (0 until expectedBeforeSync.length()).map { expectedBeforeSync.getString(it) },
            diff.missingChunks.map { it.chunkId },
        )
    }

    @Test
    fun `buildRequest matches the fixture REQUEST message`() {
        val diff = PeerSync.computeDiff(peerA, peerB, datasetId, namespace)
        val request = PeerSync.buildRequest(diff, resume = requestMessage.getBoolean("resume"))

        val expectedChunkIds = chunkIdsOf(requestMessage.getJSONArray("chunks"))
        assertEquals(expectedChunkIds, request.chunks.map { it.chunkId })
        assertEquals(requestMessage.getLong("max_total_bytes"), request.maxTotalBytes)
        assertEquals(requestMessage.getBoolean("resume"), request.resume)
    }

    @Test
    fun `after applying the request, node-a is missing nothing`() {
        val diff = PeerSync.computeDiff(peerA, peerB, datasetId, namespace)

        val merged = peerA.datasets[0].chunks.associateBy { it.chunkId }.toMutableMap()
        for (missing in diff.missingChunks) {
            merged[missing.chunkId] = ChunkSummary(missing.chunkId, missing.chunkHash, missing.sizeBytes, missing.priority)
        }
        val peerAAfterSync = peerA.copy(
            datasets = listOf(peerA.datasets[0].copy(chunks = merged.values.toList())),
        )

        val diffAfterSync = PeerSync.computeDiff(peerAAfterSync, peerB, datasetId, namespace)

        val expectedAfterSync = exchange.getJSONObject("expected").getJSONArray("a_missing_chunks_after_sync")
        assertEquals(
            (0 until expectedAfterSync.length()).map { expectedAfterSync.getString(it) },
            diffAfterSync.missingChunks.map { it.chunkId },
        )
    }

    @Test
    fun `computeDiff rejects two manifests that both claim the same dataset_version`() {
        val peerBWrongManifest = peerB.copy(
            datasets = listOf(peerB.datasets[0].copy(manifestId = "demo:official:999")),
        )

        val thrown = assertThrows(PeerSyncException::class.java) {
            PeerSync.computeDiff(peerA, peerBWrongManifest, datasetId, namespace)
        }
        assert(thrown.message?.contains("manifest mismatch") == true) {
            "expected 'manifest mismatch' in exception message, got: ${thrown.message}"
        }
    }

    @Test
    fun `computeDiff treats a peer carrying a newer manifest_id as a DTN supersession, not an error`() {
        // A walks into a shelter carrying v137; everyone there (peerB) is
        // still on v136. This must produce a usable diff, not throw.
        val newerChunk = ChunkSummary(
            chunkId = "resilientgeo-demo:chunk:137:dahu:road:000",
            chunkHash = "sha256:1111111111111111111111111111111111111111111111111111111111111111",
            sizeBytes = 512,
            priority = Priority.CRITICAL,
        )
        val peerNewerManifest = peerA.copy(
            datasets = listOf(
                peerA.datasets[0].copy(
                    manifestId = "resilientgeo-demo:manifest:137",
                    datasetVersion = 137,
                    chunks = listOf(newerChunk),
                ),
            ),
        )

        val diff = PeerSync.computeDiff(peerNewerManifest, peerB, datasetId, namespace)

        // Remote (peerB, v136) has nothing local doesn't already have --
        // local's own newer manifest_id stays current.
        assertEquals("resilientgeo-demo:manifest:137", diff.manifestId)
        assertNull(diff.supersededManifestId)
        assertEquals(emptyList<ChunkRef>(), diff.missingChunks)
        assertEquals(emptyList<ChunkRef>(), diff.staleChunks)
    }

    @Test
    fun `computeDiff requests every chunk from a peer carrying a newer manifest_id`() {
        val newerChunk = ChunkSummary(
            chunkId = "resilientgeo-demo:chunk:137:dahu:road:000",
            chunkHash = "sha256:1111111111111111111111111111111111111111111111111111111111111111",
            sizeBytes = 512,
            priority = Priority.CRITICAL,
        )
        val peerBNewerManifest = peerB.copy(
            datasets = listOf(
                peerB.datasets[0].copy(
                    manifestId = "resilientgeo-demo:manifest:137",
                    datasetVersion = 137,
                    chunks = listOf(newerChunk),
                ),
            ),
        )

        val diff = PeerSync.computeDiff(peerA, peerBNewerManifest, datasetId, namespace)

        assertEquals("resilientgeo-demo:manifest:137", diff.manifestId)
        assertEquals("resilientgeo-demo:manifest:136", diff.supersededManifestId)
        assertEquals(emptyList<ChunkRef>(), diff.staleChunks)
        assertEquals(listOf("resilientgeo-demo:chunk:137:dahu:road:000"), diff.missingChunks.map { it.chunkId })

        val request = PeerSync.buildRequest(diff)
        assertEquals("resilientgeo-demo:manifest:136", request.supersededManifestId)
    }
}
